src/status_emitter/json.rs

```diff
@@ -0,0 +1,157 @@
+use super::RevisionStyle;
+use super::StatusEmitter;
+use super::Summary;
+use super::TestStatus;
+use crate::TestOk;
+use crate::test_result::TestResult;
+
+use std::boxed::Box;
+use std::fmt::Debug;
+use std::path::{Path, PathBuf};
+
+use bstr::ByteSlice;
+
+// MAINTENANCE REGION START
+
+// When integrating with a new libtest version, update all emit_xxx functions.
+
+fn emit_suite_end(failed: usize, filtered_out: usize, ignored: usize, passed: usize, status: &String,) {
+    // Adapted from test::formatters::json::write_run_finish().
+    println!(r#"{{ "type": "suite", "event": "{status}", "passed": {passed}, "failed": {failed}, "ignored": {ignored}, "measured": 0, "filtered_out": {filtered_out} }}"#);
+}
+
+fn emit_test_end(name: &String, status: &String, error_output: &String) {
+    let escaped_name = serde_json::to_string(name).unwrap();
+    let escaped_error_output = if error_output.is_empty() {
+        String::new()
+    } else {
+        format!(r#", "stdout": {}"#, serde_json::to_string(error_output).unwrap())
+    };
+
+    // Adapted from test::formatters::json::write_event().
+    println!(r#"{{ "type": "test", "event": "{status}", "name": {escaped_name}{escaped_error_output} }}"#);
+}
+
+fn emit_test_start(name: &String) {
+    let escaped_name = serde_json::to_string(name).unwrap();
+
+    // Adapted from test::formatters::json::write_test_start().
+    println!(r#"{{ "type": "test", "event": "started", "name": {escaped_name} }}"#);
+}
+
+// MAINTENANCE REGION END
+
+/// A JSON output emitter.
+#[derive(Clone)]
+pub struct JSON;
+
+impl StatusEmitter for JSON {
+    /// Create a report about the entire test run at the end.
+    fn finalize(
+        &self,
+        failed: usize,
+        succeeded: usize,
+        ignored: usize,
+        filtered: usize,
+        aborted: bool,
+    ) -> Box<dyn Summary> {
+        let status = if aborted || failed > 0 {
+            String::from("failed")
+        } else {
+            String::from("ok")
+        };
+
+        emit_suite_end(failed, filtered, ignored, succeeded, &status);
+
+        Box::new(())
+    }
+
+    /// Invoked the moment we know a test will later be run.
+    /// Useful for progress bars and such.
+    fn register_test(&self, path: PathBuf) -> Box<dyn TestStatus + 'static> {
+        let name = path.to_str().unwrap().to_string();
+
+        emit_test_start(&name);
+
+        Box::new(JSONStatus {
+            name,
+            path,
+            revision: String::new(),
+            style: RevisionStyle::Show,
+        })
+    }
+}
+
+/// Information about a specific test run.
+pub struct JSONStatus {
+    name: String,
+    path: PathBuf,
+    revision: String,
+    style: RevisionStyle,
+}
+
+impl TestStatus for JSONStatus {
+    /// A test has finished, handle the result immediately.
+    fn done(&self, result: &TestResult, aborted: bool) {
+        let status = if aborted {
+            String::from("timeout")
```

Comment:
```suggestion
            "timeout"
```

similarly below for the other status strings

Reply:
Done.